src/client/java/net/neoforged/neoforge/client/network/registration/ClientNetworkRegistry.java

```diff
@@ -0,0 +1,272 @@
+
+    private static boolean setupClient = false;
+
+    private ClientNetworkRegistry() {}
+
+    /**
+     * Sets up the client network registry by firing {@link RegisterClientPayloadHandlersEvent}, updating the payload
+     * registrations for clientbound payloads in {@link #PAYLOAD_REGISTRATIONS}.
+     */
+    public static void setup() {
+        if (!NetworkRegistry.setup) {
+            throw new IllegalStateException("ClientNetworkRegistry cannot be set up before main NetworkRegistry");
+        }
+        if (setupClient) {
+            throw new IllegalStateException("The client network registry can only be set up once.");
+        }
+
+        ModLoader.postEvent(new RegisterClientPayloadHandlersEvent());
+
+        List<ResourceLocation> missingHandlers = PAYLOAD_REGISTRATIONS.values()
+                .stream()
+                .map(Map::values)
+                .flatMap(Collection::stream)
+                .filter(reg -> {
+                    if (!reg.matchesFlow(PacketFlow.CLIENTBOUND)) {
+                        return false;
+                    }
+
+                    for (ConnectionProtocol protocol : reg.protocols()) {
+                        if (!CLIENTBOUND_HANDLERS.get(protocol).containsKey(reg.id())) {
+                            return true;
+                        }
+                    }
+                    return false;
+                })
+                .map(PayloadRegistration::id)
+                .distinct()
+                .toList();
+        if (!missingHandlers.isEmpty()) {
+            throw new IllegalStateException("Some clientbound payloads are missing client-side handlers: " + missingHandlers);
+        }
+
+        setupClient = true;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends CustomPacketPayload> void register(CustomPacketPayload.Type<T> type, HandlerThread thread, IPayloadHandler<T> handler) {
+        if (setupClient) {
+            throw new UnsupportedOperationException("Cannot register client-side handler for payload " + type.id() + " after registration phase.");
+        }
+
+        if (thread == HandlerThread.MAIN) {
+            handler = new MainThreadPayloadHandler<>(handler);
+        }
+
+        boolean found = false;
+        for (var entry : PAYLOAD_REGISTRATIONS.entrySet()) {
+            ConnectionProtocol protocol = entry.getKey();
+            Map<ResourceLocation, PayloadRegistration<?>> registrations = entry.getValue();
+
+            PayloadRegistration<T> registration = (PayloadRegistration<T>) registrations.get(type.id());
+            if (registration == null) {
+                continue;
+            }
+
+            if (!registration.matchesFlow(PacketFlow.CLIENTBOUND)) {
+                throw new IllegalArgumentException("Cannot register client handler for serverbound payload " + type);
+            }
+
+            found = true;
+
+            registerHandler(CLIENTBOUND_HANDLERS, protocol, PacketFlow.CLIENTBOUND, type, handler);
```

Comment:
Where's this `CLIENTBOUND_HANDLERS` field defined? Shouldn't it be in this class?